cdap-app-fabric/src/main/java/org/apache/twill/internal/yarn/AbstractYarnAMClient.java

```diff
@@ -0,0 +1,301 @@
+
+  // Map from a unique ID to inflight requests
+  private final Multimap<String, T> containerRequests;
+
+  // List of requests pending to send through allocate call
+  private final List<T> requests;
+  // List of requests pending to remove through allocate call
+  private final List<T> removes;
+  //List of pending blacklist additions for the next allocate call
+  private final List<String> blacklistAdditions;
+  //List of pending blacklist removals for the next allocate call
+  private final List<String> blacklistRemovals;
+  //Keep track of blacklisted resources
+  private final List<String> blacklistedResources;
+  /**
+   * Contains a list of known unsupported features.
+   */
+  protected final Set<String> unsupportedFeatures = Sets.newHashSet();
+
+  protected final ContainerId containerId;
+  protected InetSocketAddress trackerAddr;
+  protected URL trackerUrl;
+
+  /**
+   * Constructs an instance of AMClient.
+   *
+   * @param containerIdEnvName Name of the environment variable that contains value of the AM container ID.
+   */
+  protected AbstractYarnAMClient(String containerIdEnvName) {
+    String masterContainerId = System.getenv().get(containerIdEnvName);
+    Preconditions.checkArgument(masterContainerId != null,
+                                "Missing %s from environment", containerIdEnvName);
+    this.containerId = ConverterUtils.toContainerId(masterContainerId);
+    this.containerRequests = ArrayListMultimap.create();
+    this.requests = Lists.newLinkedList();
+    this.removes = Lists.newLinkedList();
+    this.blacklistAdditions = Lists.newArrayList();
+    this.blacklistRemovals = Lists.newArrayList();
+    this.blacklistedResources = Lists.newArrayList();
+  }
+
+
+  @Override
+  public final ContainerId getContainerId() {
+    return containerId;
+  }
+
+  @Override
+  public final void setTracker(InetSocketAddress trackerAddr, URL trackerUrl) {
+    this.trackerAddr = trackerAddr;
+    this.trackerUrl = trackerUrl;
+  }
+
+  @Override
+  public final synchronized void allocate(float progress, AllocateHandler handler) throws Exception {
+    // In one allocate cycle, either only do new container request or removal of requests.
+    // This is a workaround for YARN-314.
+    // When remove a container request, AMRMClient will send a container request with size = 0
+    // With bug YARN-314, if we mix the allocate call with new container request of the same priority,
+    // in some cases the RM would not see the new request (based on sorting of resource capability),
+    // but rather only see the one with size = 0.
+    if (removes.isEmpty()) {
+      for (T request : requests) {
+        addContainerRequest(request);
+      }
+      requests.clear();
+    } else {
+      for (T request : removes) {
+        try {
+          removeContainerRequest(request);
+        } catch (NullPointerException e) {
+          // workaround for TWILL-186
+          LOG.warn("Got exception.", e);
```

Comment:
Log it as info or even debug as we know the cause and don't want to get user panic.